libftl/logging.c

```diff
@@ -32,7 +32,7 @@ void ftl_log_msg(ftl_stream_configuration_private_t *ftl, ftl_log_severity_t log
 
   m.msg.log.log_level = log_level;
   va_start(args, fmt);
-  vsnprintf(m.msg.log.string, sizeof(m.msg.log.string), fmt, args);
+  vsnprintf_s(m.msg.log.string, sizeof(m.msg.log.string), sizeof(m.msg.log.string), fmt, args);
```

Comment:
I believe you want something like strlen here, sizeof will return the size of the type which I think is char*, so 4 bytes.

Reply:
I'm not much of a C guy (more of a C++ heathen), but I'd always refer to the standard in such cases:
http://www.open-std.org/jtc1/sc22/WG14/www/docs/n1256.pdf page 92, 6.5.3.4 3rd paragraph

And voila, like C++, C's arrays degrade into pointers when passed to a function that accepts pointers, but the sizeof operator actually returns the full size of the array.
The standard even gives you a neat example covering your use case *g*

Btw, according to wikipedia where it is linked, that is the "Final version of the C99 standard with corrigenda TC1, TC2, and TC3 included, formatted as a draft".
If that standardization process is anything like C++'s, that's about as close as you can get to the real standard without paying money :)

Reply:
strlen now used in the next iteration.

Reply:
thnkas guys for the tips, I've been living in c# world for a long time 😄

Reply:
Just dropping in here, sorry for the confusion.
The type of m.msg.log.string is char[]. It's an allocated array in the struct and has a size. sizeof(char[]) is correct here and will give the number of bytes in the total array. strlen() is incorrect here as the array isn't yet initialized. It will just search through uninitialized memory until it happens to find a 0.

@TomyLobo I think you're misinterpreting the standard. Footnote 88 on page 80 under 6.5.3.4 explains how sizeof() applies to a parameter degraded array, which is that it returns the size of the pointer type, not the whole array.

Reply:
There is however an issue with the second sizeof() parameter (the count parameter). You either need to subtract 1 to allow for the null terminator and add error handling code to check if vsnprintf_s failed to write the whole string, or since this is just for logging, you can just replace it with _TRUNCATE and let it automatically truncate the string.

Reply:
I was confused for a while there.
Apparently, pdf page 92 is standard page 80 :)

Anyway, I actually have no idea what type the "string" field has. I assumed that, being a field it was not char[], but char[x] (which behave differently in sizeof, at least in C++)

So, what is it? char[] or char[x]

Reply:
Ah, sorry about the page confusion.
The "string" variable is a char[1024] in a struct which is in the msg union, which is in the m struct.
```
typedef struct {
  int log_level;
  char string[1024];
}ftl_status_log_msg_t;
```

Reply:
https://ideone.com/1YQUTV there we go
So this means that "m.msg.log.string" cannot actually be a char[], assuming this piece of code compiles in gcc 6.3 :)

This means it's likely char[x] and sizeof was correct.
If, however, it is char*, you cannot use sizeof. In that case, you need the size from somewhere else. If it's not initialized and you did not [store the length somewhere alongside the pointer](https://en.cppreference.com/w/cpp/string/basic_string), good luck figuring out the length :)

Reply:
Yup so seems like this is a char[x], so I'll move this back to sizeof.
And for the count we'll use the _TRUNCATE for string truncation.

Reply:
Good call. Thanks everyone!